compiler/rustc_mir/src/borrow_check/diagnostics/mutability_errors.rs

```diff
@@ -463,6 +466,63 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
         err.buffer(&mut self.errors_buffer);
     }
 
+    fn suggest_similar_mut_method_for_for_loop(&self, err: &mut DiagnosticBuilder<'_>) {
+        let hir = self.infcx.tcx.hir();
+        let mir_hir_id = self.mir_hir_id();
+        let node = hir.item(mir_hir_id);
+        if let hir::ItemKind::Fn(_, _, body_id) = node.kind {
+            let body = hir.body(body_id);
+            if let hir::ExprKind::Block(block, _) = body.value.kind {
+                if let Some(expr) = block.expr {
+                    if let hir::ExprKind::DropTemps(expr) = expr.kind {
+                        if let hir::ExprKind::Match(expr, ..) = expr.kind {
+                            if let hir::ExprKind::Call(_, [expr, ..]) = expr.kind {
+                                if let hir::ExprKind::MethodCall(path_segment, ..) = expr.kind {
+                                    if let Some(hir_id) = path_segment.hir_id {
```

Comment:
I think it's time to use the `if_chain` crate in the compiler instead of just in clippy :laughing:

Reply:
Heh, I did think about CCing you when I saw that xD

When playing around with it I came up with this. Still gnarly but at least it separates the actual code from taking the structure apart.
```rust
fn suggest_similar_mut_method_for_for_loop(&self, err: &mut DiagnosticBuilder<'_>) {
        let hir = self.infcx.tcx.hir();
        let node = hir.item(self.mir_hir_id());
        use hir::{
            Expr,
            ExprKind::{Block, Call, DropTemps, Match, MethodCall},
        };
        if let hir::ItemKind::Fn(_, _, body_id) = node.kind {
            if let Block(
                hir::Block {
                    expr:
                        Some(Expr {
                            kind:
                                DropTemps(Expr {
                                    kind:
                                        Match(
                                            Expr {
                                                kind:
                                                    Call(
                                                        _,
                                                        [Expr {
                                                            kind: MethodCall(path_segment, ..),
                                                            hir_id,
                                                            ..
                                                        }, ..],
                                                    ),
                                                ..
                                            },
                                            ..,
                                        ),
                                    ..
                                }),
                            ..
                        }),
                    ..
                },
                _,
            ) = hir.body(body_id).value.kind
            {
                let opt_suggestions = path_segment
                    .hir_id
                    .map(|path_hir_id| self.infcx.tcx.typeck(path_hir_id.owner))
                    .and_then(|typeck| typeck.type_dependent_def_id(*hir_id))
                    .and_then(|def_id| self.infcx.tcx.impl_of_method(def_id))
                    .map(|def_id| self.infcx.tcx.associated_items(def_id))
                    .map(|assoc_items| {
                        assoc_items
                            .in_definition_order()
                            .map(|assoc_item_def| assoc_item_def.ident)
                            .filter(|&ident| {
                                let original_method_ident = path_segment.ident;
                                original_method_ident != ident
                                    && ident
                                        .as_str()
                                        .starts_with(&original_method_ident.name.to_string())
                            })
                            .map(|ident| format!("{}()", ident))
                    });

                if let Some(suggestions) = opt_suggestions {
                    err.span_suggestions(
                        path_segment.ident.span,
                        &format!("use mutable method"),
                        suggestions,
                        Applicability::MaybeIncorrect,
                    );
                }
            }
        }
    }
```